client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchRequest.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+package org.elasticsearch.client.asyncsearch;
+
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.client.Validatable;
+import org.elasticsearch.common.unit.TimeValue;
+
+import java.util.Objects;
+
+/**
+ * A request to track asynchronously the progress of a search against one or more indices.
+ */
+public class SubmitAsyncSearchRequest implements Validatable {
+
+    private TimeValue waitForCompletion;
+    private Boolean cleanOnCompletion;
+    private TimeValue keepAlive;
+    private final SearchRequest searchRequest;
+
+    /**
+     * Create a request to submit an async search.
+     * Target indices, queries and all other search related options should be set on
+     * the input {@link SearchRequest}.
+     * @param searchRequest the actual search request to submit
+     */
+    public SubmitAsyncSearchRequest(SearchRequest searchRequest) {
```

Comment:
The original request takes the indices to search in the constructor rather than the `SearchRequest` that is created internally. Can you adapt it to match the behavior of the transport request ?

Reply:
I thought about it but decided against it. The transport requests signature ` public SubmitAsyncSearchRequest(SearchSourceBuilder source, String... indices)` doesn't allow setting things in the SearchRequest thats constructed internally, like e.g. `search_type` or `routing`. 
You probably would need to do sth. like this on the client side which I find cumbersome:
```
SearchSourceBuilder sourceBuilder = new SearchSourceBuilder()....;        
SubmitAsyncSearchRequest submitRequest = new SubmitAsyncSearchRequest(indices, sourceBuilder);
submitRequest.getSearchRequest().routing(). 
```
I think on the client side its easier for the user to pass in the full `SearchRequest` rather than having to do things like this. Wdyt?

Reply:
Then maybe we should expose the search request options that we allow to modify directly in the `SubmitAsyncSearchRequest` and hide the internal `searchRequest` ? Whatever we do here I'd prefer to have the same signature between hlrc and transport since they should be equivalent.

Reply:
> here I'd prefer to have the same signature between hlrc and transport since they should be equivalent.

I'm open to adding a bunch of setters to the client side SubmitAsyncSearchRequest but I wonder if its really necessary. Why do the transport and the client side request ctors need to be equivalent? With the transport client gone I only see the transport request used internally, everything else would go through the RestSubmitAsyncSearchAction which handles creating the transport request, or am I missing something? Maybe @javanna has some ideas here as well.

Reply:
I agree but in an ideal world we'd have a single class here so I prefer if we don't diverge in the hlrc. Exposing the search request directly is tricky since some of the options are not allowed (scroll, ccs) and we change the default internally. Hiding the search request entirely behind direct setters and getters seems safer and would avoid the confusion when building an async search request ?